libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -128,238 +169,771 @@ BeaconSetupService::State BeaconSetupService::OnIdle()
+    return State::WAIT_FOR_SHARE;
+  }
+
+  state_machine_->Delay(std::chrono::milliseconds(100));
+  FETCH_LOG_INFO(LOGGING_NAME, "Waiting for all peers to be ready before starting DKG. Ready: ",
+                 ready_connections_.size(), " expect: ", beacon_->aeon.members.size() - 1);
+  return State::WAIT_FOR_READY_CONNECTIONS;
 }
 
-BeaconSetupService::State BeaconSetupService::WaitForIDs()
+BeaconSetupService::State BeaconSetupService::OnWaitForShares()
 {
-  std::lock_guard<std::mutex> lock(mutex_);
-  assert(beacon_ != nullptr);
+  std::unique_lock<std::mutex> lock{mutex_};
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_SHARE));
+
+  if ((coefficients_received_.size() == beacon_->aeon.members.size() - 1) &&
+      (shares_received_.size() == beacon_->aeon.members.size() - 1))
+  {
+    lock.unlock();
+    BroadcastComplaints();
+
+    // Clean up
+    coefficients_received_.clear();
+    shares_received_.clear();
+
+    return State::WAIT_FOR_COMPLAINTS;
+  }
+
+  state_machine_->Delay(std::chrono::milliseconds(10));
+  return State::WAIT_FOR_SHARE;
+}
+
+BeaconSetupService::State BeaconSetupService::OnWaitForComplaints()
+{
+  std::unique_lock<std::mutex> lock{mutex_};
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_COMPLAINTS));
+
+  if (complaints_manager_.IsFinished(beacon_->manager.polynomial_degree()))
+  {
+    // Complaints at this point consist only of parties which have received over threshold number of
+    // complaints
+    FETCH_LOG_INFO(LOGGING_NAME, "Node ", beacon_->manager.cabinet_index(), " complaints size ",
+                   complaints_manager_.Complaints().size());
+    complaints_answer_manager_.Init(complaints_manager_.Complaints());
+    lock.unlock();
+    BroadcastComplaintsAnswer();
+
+    return State::WAIT_FOR_COMPLAINT_ANSWERS;
+  }
+
+  state_machine_->Delay(std::chrono::milliseconds(10));
+  return State::WAIT_FOR_COMPLAINTS;
+}
+
+BeaconSetupService::State BeaconSetupService::OnWaitForComplaintAnswers()
+{
+  std::unique_lock<std::mutex> lock{mutex_};
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_COMPLAINT_ANSWERS));
 
-  // Checking incoming messages
-  std::vector<CabinetMemberDetails> remaining;
-  for (auto &member : member_details_queue_)
+  if (complaints_answer_manager_.IsFinished())
   {
-    auto it = beacon_->aeon.members.find(member.identity);
-    if (it == beacon_->aeon.members.end())
+    lock.unlock();
+    if (BuildQual())
     {
-      remaining.push_back(member);
+      FETCH_LOG_INFO(LOGGING_NAME, "Node ", beacon_->manager.cabinet_index(), "build qual size ",
+                     beacon_->manager.qual().size());
+      beacon_->manager.ComputeSecretShare();
+      BroadcastQualCoefficients();
+
+      complaints_manager_.Clear();
+      return State::WAIT_FOR_QUAL_SHARES;
     }
     else
     {
-      member_details_[member.identity] = member.id;
+      complaints_manager_.Clear();
+      // TODO(jmw): procedure failed for this node
```

Comment:
I think it might be a good idea to reset the whole thing if minimum conditions are not met - chat tomorrow?

Reply:
Sure. However, this is just a single node failing and therefore does not require the DKG being reset. However, there are points of failure where everyone fails and that will require a reset, which is currently not handled.